CPT: Render basic post details in custom post type listing

aduth opened the pull request:
This pull request seeks to continue iterating on the layout of the custom post type post listing screen, adding featured image, last edited date, a view button, and edit link to each post list entry.

Before|After
---|---
![Before](https://cloud.githubusercontent.com/assets/1779930/13787063/bdbc65ca-eab0-11e5-9337-efc3281cc011.png)|![After](https://cloud.githubusercontent.com/assets/1779930/13786827/d6048f96-eaaf-11e5-9e61-81dfa9e370dc.png)

__Note:__ The layout is very much a work-in-progress and will continue to be refined over time.

__Reference design:__ https://cloudup.com/cdHFJ0ddwl7

__Implementation notes:__

- `getSiteSlug` must have been reimplemented, as while it does currently exist on the decorated `site` object, this is non-ideal and calculated values will be removed from the Redux store site state in the near future. See [`Site.prototype.updateComputedAttributes`](https://github.com/Automattic/wp-calypso/blob/3bde0a6b63d087b31cc4d026fe52b67c03bb0a27/client/lib/site/index.js#L85-L116) for the reference implementation and #2757 as the tracking issue for site state cleanup.
- `<PostTypePost />` will likely be split into separate components in a future iteration as each piece of the layout continues to be fleshed out.

__Testing instructions:__

1. Navigate to My Sites
2. If the current site does not have a custom post type enabled (e.g. Testimonials, Portfolio), switch to a site which has at least one enabled
3. Click the custom post type navigation item in the sidebar
4. Note that after posts have loaded, the layout reflects the "After" screenshot above for any posts which exist
5. Note that the edit link (title), view link, updated date, and featured image are accurate

Caveats:

- There is no loading indicator for posts
- There is no warning shown if no posts exist, only a blank page

nylen commented on client/state/sites/selectors.js:
```diff
@@ -10,3 +10,23 @@
+	}
+
+	return site.URL.replace( /^https?:\/\//, '' ).replace( /\//g, '::' );
+}
```
This differs from the current implementation in that `hasConflict` is not included:  https://github.com/Automattic/wp-calypso/search?q=hasConflict

More background at 5836-gh-calypso-pre-oss - looks like this is intended to handle fairly extreme edge cases around Jetpack and WP.com sites ending up with the same URL, and the current implementation would need some changes to fit with a reduxified sites list.  I personally think it would be ok to punt on it but wanted to mention.

aduth replied:
>This differs from the current implementation in that `hasConflict` is not included:  https://github.com/Automattic/wp-calypso/search?q=hasConflict

Good eyes. I must have had my blinders on when I had overlooked that part of the condition ( :see_no_evil: ).

@enejb : There's quite a bit of background here. Do you think this is logic that needs to be preserved moving forward in the sites state? Apologies to shine light on this again, but I'm finding it difficult to find the conclusion of the previous discussion.

enejb replied:
@aduth without it some users will not be able to access the site that they want. `hasConflict` is set when it sees that a .com site has the same url as a jetpack site. It sets the primary domain of the .com site to the `something.wordpress.com` even the site is domain mapped for example. This allow the user to navigate calypso without conflicts. 

Since otherwise the user would not be able to land on the site that they want. Please keep it otherwise I am sure we will get someone telling us that they can't access the site.

nylen commented on client/state/ui/editor/selectors.js:
```diff
@@ -27,3 +38,35 @@ export function isEditorNewPost( state ) {
+	switch ( type ) {
+		case 'post': path = '/post'; break;
+		case 'page': path = '/page'; break;
+		default: path = `/type/${ type }`; break;
```
I think `/edit/${ type }` is a better choice here.

nylen replied:
Perhaps also `/view/${ type }` rather than `/types/${ type }` for the listing page?

aduth replied:
>I think `/edit/${ type }` is a better choice here.

Ignoring everything else, I'd agree that `/edit/${ type }` is a nice path. It does lose the consistency we have with `/posts` -> `/post` and `/pages` -> `/page`. I'm still not totally content with the `/types` path either at this point (#3689).

>Perhaps also `/view/${ type }` rather than `/types/${ type }` for the listing page?

"View" seems like a pretty loaded term that'd be ripe for conflict, and it's not immediately obvious it relates to posts (vs. something else, customizer, theme preview, etc). In any case, `/list` would probably be more appropriate (since it's not viewing a single item), but suffers from similar concerns.

For reference, the wp-admin paths are:
- List view: `/edit.php?post_type=jetpack-testimonial`
- Detail view: `/edit.php?post=1&action=edit`

I don't particularly like these paths, but worth highlighting nonetheless.

We're still early enough in the process that we can iterate on the path, especially as more people come to find the page and have their own opinions.

nylen replied:
I agree `/view` could be a bit ambiguous - I think `/list` is good though.

aduth replied:
>I agree `/view` could be a bit ambiguous - I think `/list` is good though.

I'll plan to create a follow-up pull request specifically for starting a dialogue around these paths, as I'd prefer to avoid it holding up this one.

nylen commented:
Works as intended.

One thing worth noting, but I doubt we can do anything about it, is that featured images still show up in the listing even if the theme doesn't do anything with them:

![image](https://cloud.githubusercontent.com/assets/227022/13836765/9f4c51a4-ebd2-11e5-9374-0cfd224c5014.png)

<a href="https://jnylen0.wordpress.com/testimonial/"><img src="https://cloud.githubusercontent.com/assets/227022/13836748/792a1f24-ebd2-11e5-95be-c8b0063c1253.png"></a>

aduth commented:
@nylen : Interesting, [the theme's](https://wordpress.org/themes/sunspot/) `functions.php` file does add support for `post-thumbnails`, and the `/post-types` endpoint for your site reports that the type does support thumbnails, though the `/sites/%s` endpoint claims that `options.featured_images_enabled` is `false`. Clearly the images aren't shown on the front-end of the site, but also curiously can be assigned in the wp-admin editor.

I've opened #4124 as a follow-up task for this, as there's much more to be fleshed out with regard to thumbnails in any case.

aduth commented:
Rebased and added bba3d2c to account for site conflicts, including two new site state selectors `getSiteCollisions( state )` and `isSiteConflicting( state, siteId )`. See also [reference implementation](https://github.com/Automattic/wp-calypso/blob/f43de1023ad62348d4fd596c5413a0fe54eca2b3/client/lib/sites-list/list.js#L149-L169).

aduth commented:
@enejb : Would you mind reviewing the latest changes in bba3d2c to confirm that it covers the collisions needs well?

aduth commented:
@enejb : Good point, and one that doesn't seem to be accounted for in our current `sites-list` implementation. I've pushed d72d607 to ensure it's considered here.

enejb commented:
Thanks for the fix!
